pkg/pillar/cmd/zedagent/handleconfig.go

```diff
@@ -150,6 +152,104 @@ var nilUUID uuid.UUID
 // current epoch received from controller
 var controllerEpoch int64
 
+type configSource int
+
+const (
+	fromController configSource = iota
+	savedConfig
+	fromBootstrap
+)
+
+func loadBootstrapConfig(getconfigCtx *getconfigContext) {
+	//  Check if bootstrap config has been already loaded.
+	if _, err := os.Stat(types.BootstrapConfFileName); err != nil {
+		if os.IsNotExist(err) {
+			// No bootstrap config to read
+			return
+		}
+		// Potentially there is a problem reading bootstrap config,
+		// but continue and try anyway.
+		log.Errorf("Failed to stat bootstrap config: %v", err)
+	}
+	changed, configSha, err := fileutils.CompareSha(
+		types.BootstrapConfFileName, types.BootstrapShaFileName)
+	if err != nil {
+		log.Errorf("CompareSha failed for bootstrap config: %s", err)
+		// We will not record SHA for applied bootstrap config
+		// and as a result load it again with the next boot.
+		// However, with config timestamp preventing accidental revert
+		// to an older configuration, this should not be a problem.
+		configSha = nil
+	} else if !changed {
+		// This bootstrap config was already applied.
+		return
+	}
+
+	// Load file content.
+	contents, err := ioutil.ReadFile(types.BootstrapConfFileName)
+	if err != nil {
+		log.Errorf("Failed to read bootstrap config: %v", err)
+		return
+	}
+
+	// Mark bootstrap config as processed by storing SHA hash of its content
+	// under /persit/ingested.
+	// We do this even even if the unmarshalling (or anything else) below fails.
+	// This is to avoid repeated failing attempts to load invalid config on each boot.
+	defer func() {
+		if configSha != nil {
+			err := fileutils.SaveShaInFile(types.BootstrapShaFileName, configSha)
+			if err != nil {
+				log.Errorf("Failed to save SHA of bootstrap config: %v", err)
+			}
+		}
+	}()
+
+	// Unmarshal BootstrapConfig.
+	bootstrap := zconfig.BootstrapConfig{}
+	err = proto.Unmarshal(contents, &bootstrap)
+	if err != nil {
+		log.Errorf("Failed to unmarshal bootstrap config: %v", err)
```

Comment:
I wonder if we should make this and subsequent errors in this function result in some >10 LED blinking pattern.

Reply:
Done